### DISCUSSION TITLE
Express custom server not working

### DISCUSSION THREAD

**Discussion by im-ghost** [vercel/next.js#37896] [Help] [answered]
I tried using express custom server for my next project based and the example given,but I get an error saying next is not a function

**Comment by icyJoseph** [comment]
Got a repo to look at?

**Reply by im-ghost** [reply]
https://github.com/Peculiar-codes/chat-app.git



The error branch sir

**Reply by im-ghost** [reply]
https://github.com/Peculiar-codes/chat-app/tree/error

**Reply by ugursepetci** [reply]
project not public

> https://github.com/Peculiar-codes/chat-app/tree/error

**Reply by im-ghost** [reply]
> 

Already public

**Reply by ugursepetci** [reply]
Can you replace to express.ts?
https://github.com/vercel/next.js/blob/canary/examples/custom-server-express/server.js

**Reply by im-ghost** [reply]
I don get you sir, should I replace pages/api/server.ts to pages/api/express.ts?

**Reply by ugursepetci** [reply]
no it shouldn't be in pages/api.

It should be named server.js in the main folder. and you have to install 'express' as package.

can you inspect to https://github.com/vercel/next.js/tree/canary/examples/custom-server-express

**Comment by icyJoseph** [comment, marked as answer]
Why is your server inside `pages/api`? It should be at the root. 

https://nextjs.org/docs/advanced-features/custom-server

Move `pages/api/server.ts` to the root of the project, and run it from there. I am not sure if it can be a TypeScript file though, perhaps change it to `.js`.

**Reply by im-ghost** [reply]
Thanks a lot
But am confused about something,am I supposed to run the server separately

**Reply by icyJoseph** [reply]
Yes, because you are not using the next server. You'd need to run it as `node server.js` either manually or behind a `package.json` script. 

Let me ask you something though. What is it that you are trying to achieve? Next.js is already a server, and needs no express integration for the vast majority of use cases. The API routes are already running Node.js and can process requests on demand.

Stubbing an entire express server into one of them would be wasteful, the api route is destroyed if inactive, or if too much traffic is incoming, more instances are created. Why would you have a server there, if it's already a server of its own.

**Reply by im-ghost** [reply]
Am trying to use socket.io sir,and I learnt it doesn't work with nextjs

**Reply by im-ghost** [reply]
The server and the next must be on different port,right?

**Reply by icyJoseph** [reply]
I think you should probably do as shown here: https://wallis.dev/blog/socketio-with-nextjs-and-es6-import

It combines the same server to offer both the next.js app service and the socket-io service.

**Reply by im-ghost** [reply]
Thanks a lot

**Reply by im-ghost** [reply]
From what I understand from the tutorial, it still going to be a different server

**Reply by icyJoseph** [reply]
Yeah, as far as I have been reading my notes, and online, a custom Next.js server is the only way. 

I have used a solution called Pusher with Next.js, but that's like, a server with extra steps and could cost money.

It's one of the pains of serverless/lambda architectures.

**Reply by im-ghost** [reply]
Ooh,Have heard of pusher.
But socket.io-client doesn't seem to work in next, can't emit events
